rand_core/src/lib.rs

```diff
@@ -247,55 +247,138 @@ pub enum ErrorKind {
     __Nonexhaustive,
 }
 
-#[cfg(feature="std")]
+impl ErrorKind {
+    pub fn description(&self) -> &'static str {
+        match *self {
+            ErrorKind::Unavailable => "permanent failure or unavailable",
+            ErrorKind::Transient => "transient failure",
+            ErrorKind::NotReady => "not ready yet",
+            ErrorKind::Other => "uncategorised error",
+            ErrorKind::__Nonexhaustive => unreachable!(),
+        }
+    }
+}
+
 #[derive(Debug)]
-pub struct Error {
-    pub kind: ErrorKind,
-    pub cause: Option<Box<std::error::Error>>,
+enum ErrorDetails {
+    None,
+    Str(&'static str),
+    #[cfg(feature="std")]
+    Error(Box<std::error::Error + Send + Sync>),
 }
 
-#[cfg(not(feature="std"))]
-#[derive(Debug, Clone, PartialEq, Eq)]
+impl fmt::Display for ErrorDetails {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ::std::error::Error;
+        write!(f, "{}", self.description())
+    }
+}
+
+// Impl is only used for ErrorDetails::Str case, but can't implement for an enum case
+#[cfg(feature="std")]
+impl ::std::error::Error for ErrorDetails {
+    fn description(&self) -> &str {
+        match *self {
+            ErrorDetails::None => "NA", // case shouldn't be used
+            ErrorDetails::Str(ref s) => s,
+            ErrorDetails::Error(ref e) => e.description(),
+        }
+    }
+    
+    // no cause: Error case gets unwrapped by impl for Error
+}
+
+/// Error type of random number generators
+/// 
+/// This embeds two things, an `ErrorKind`, which can be matched over, and
+/// optional details provided to `new_str` or `new_err`.
+#[derive(Debug)]
 pub struct Error {
+    /// Error kind
     pub kind: ErrorKind,
-    pub cause: Option<&'static str>,
+    details: ErrorDetails,
 }
 
 impl Error {
+    /// Create a new instance, with no details.
+    pub fn new(kind: ErrorKind) -> Self {
+        Self { kind, details: ErrorDetails::None }
+    }
+    
+    /// Create a new instance, with a static string describing the details.
+    /// 
+    /// This may be used in both `std` and `no_std` enviroments.
+    pub fn new_str(kind: ErrorKind, details: &'static str) -> Self {
+        Self { kind, details: ErrorDetails::Str(details) }
+    }
+    
+    /// Create a new instance, with a chained `Error` cause.
     #[cfg(feature="std")]
-    pub fn new(kind: ErrorKind, cause: Option<Box<std::error::Error>>) -> Error {
-        Error {
-            kind: kind,
-            cause: cause,
+    pub fn new_err<E>(kind: ErrorKind, cause: E) -> Self
+        where E: Into<Box<std::error::Error + Send + Sync>>
+    {
+        Self { kind, details: ErrorDetails::Error(cause.into()) }
+    }
+    
+    /// Get a description of the details (from str or description of chained
+    /// error).
+    /// 
+    /// In the case of a chained error, the actual error may be obtained via
+    /// `std::error::Error::cause`.
+    pub fn details(&self) -> Option<&str> {
```

Comment:
Maybe "will be optained"? Also not `description`, but `cause`.

Reply:
Um, actually I think my description is correct, but I'll try to make it clearer.

Reply:
This is the `details()` function. Anyway, the terms *are* a bit confused, although `description` should be distinct from `details`/`cause`